router/src/lib.rs

```diff
@@ -20,6 +30,28 @@ pub struct HubModelInfo {
     pub pipeline_tag: Option<String>,
 }
 
+#[derive(Clone, Deserialize)]
+pub struct HubTokenizerConfig {
+    #[serde(default)]
+    pub chat_template: Option<String>,
+}
+
+impl HubTokenizerConfig {
+    /// Apply the chat template to the chat request
+    pub(crate) fn apply_chat_template(
+        &self,
+        chat: ChatRequest,
+    ) -> Result<String, minijinja::Error> {
+        let mut env = minijinja::Environment::new();
+        let chat_template = self
+            .chat_template
+            .as_ref()
+            .ok_or(minijinja::ErrorKind::TemplateNotFound)?;
+        env.add_template("_", chat_template)?;
+        env.get_template("_")?.render(chat)
```

Comment:
@OlivierDehaene great points, I agree we should only initialize the template a single time 👍

the latest commit removes the `tokenizer_config` from the `Infer` struct in favor of a `template`.

```rust
/// Inference struct
#[derive(Clone)]
pub struct Infer {
    /// Validation
    validation: Validation,
    /// Request queue
    queue: Queue,
    /// Shared state
    shared: Arc<Shared>,
    /// Inference limit
    limit_concurrent_requests: Arc<Semaphore>,
    /// Chat template
    template: Option<Template<'static, 'static>>,
}
```

I attempted to use the `Expression` type from minijinja but it only parses single expression and not the full chat_template (for example `%`  seem to be related to templates and not expressions), so I've used a `Template` which can parse.

Note** In order to keep the `Template` alive without adding lifetimes to `Infer` (many place in TGI) I've expanded the lifetimes of the `template_str` and `env` with `Box:leak` since they will be static for the runtime of the program and can be treated as read only variables. This supports initializing the `Template` a single time instead of on each request; however is a bit of a strange pattern... 

Please let me know if theres a better approach!

Reply:
Leaking seems fine to me given the nature of this object (it lives for the duration of the program).

Reply:
Perfect